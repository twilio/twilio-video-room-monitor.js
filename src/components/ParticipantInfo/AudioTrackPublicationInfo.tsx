import React from 'react';
import styled from 'styled-components';
import {
  LocalAudioTrack,
  LocalAudioTrackPublication,
  LocalAudioTrackStats,
  RemoteAudioTrack,
  RemoteAudioTrackPublication,
  RemoteAudioTrackStats,
} from 'twilio-video';
import useIsTrackEnabled from '../../hooks/useIsTrackEnabled/useIsTrackEnabled';
import { useTrackBandwidth, useTrackData } from '../../hooks/useStats/useStats';
import useTrack from '../../hooks/useTrack/useTrack';
import { theme } from '../theme';
import Datum from '../typography/Datum/Datum';

export const AudioTrackInfo: React.FC<{
  track: LocalAudioTrack | RemoteAudioTrack;
  trackSid: string; // Passing trackSid from the publication object because it not on the LocalAudioTrack object
}> = ({ track, trackSid }) => {
  const isEnabled = useIsTrackEnabled(track);
  const trackBandwidth = useTrackBandwidth(trackSid);
  const trackData = useTrackData(trackSid) as LocalAudioTrackStats | RemoteAudioTrackStats | null;

  return (
    <>
      <Datum label="isEnabled" value={String(isEnabled)} />
      <Datum label="Bandwidth" value={String(trackBandwidth) + 'kbps'} />
      {trackData && (
        <>
          <Datum label="Codec" value={String(trackData.codec)} />
          <Datum label="Jitter" value={String(trackData.jitter)} />
          <Datum label="Packets Lost" value={String(trackData.packetsLost)} />
        </>
      )}
    </>
  );
};

const Container = styled.div`
  &:not(:last-child) {
    border-bottom: 1px solid ${theme.borderColor};
    margin-bottom: 3px;
  }
  padding-bottom: 3px;
`;

export const AudioTrackPublicationInfo: React.FC<{
  publication: LocalAudioTrackPublication | RemoteAudioTrackPublication;
}> = ({ publication }) => {
  const track = useTrack(publication) as LocalAudioTrack | RemoteAudioTrack | undefined;

  return (
    <Container>
      <Datum label="Name" value={publication.trackName} />
      <Datum label="SID" value={publication.trackSid} />
      <Datum label="isSubscribed" value={String(!!track)} />
      {track && <AudioTrackInfo track={track} trackSid={publication.trackSid} />}
    </Container>
  );
};
