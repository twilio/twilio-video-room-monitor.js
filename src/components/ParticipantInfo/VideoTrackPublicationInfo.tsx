import React from 'react';
import styled from 'styled-components';
import {
  LocalVideoTrack,
  LocalVideoTrackPublication,
  LocalVideoTrackStats,
  RemoteVideoTrack,
  RemoteVideoTrackPublication,
  RemoteVideoTrackStats,
  VideoTrack,
} from 'twilio-video';
import useIsTrackEnabled from '../../hooks/useIsTrackEnabled/useIsTrackEnabled';
import useIsTrackSwitchedOff from '../../hooks/useIsTrackSwitchedOff/useIsTrackSwitchedOff';
import usePublishPriority from '../../hooks/usePublishPriority/usePublishPriority';
import { useTrackBandwidth, useTrackData } from '../../hooks/useStats/useStatsUtils';
import useTrack from '../../hooks/useTrack/useTrack';
import useVideoTrackDimensions from '../../hooks/useVideoTrackDimensions/useVideoTrackDimensions';
import { theme } from '../theme';
import Datum from '../typography/Datum/Datum';

const getDimensionString = (dimensions?: VideoTrack.Dimensions) =>
  dimensions ? `${dimensions.width} x ${dimensions.height}` : 'undefined';

export const VideoTrackInfo: React.FC<{
  track: LocalVideoTrack | RemoteVideoTrack;
  trackSid: string; // Passing trackSid from the publication object because it not on the LocalVideoTrack object
}> = ({ track, trackSid }) => {
  const dimensions = useVideoTrackDimensions(track);
  const isSwitchedOff = useIsTrackSwitchedOff(track);
  const isEnabled = useIsTrackEnabled(track);
  const trackBandwidth = useTrackBandwidth(trackSid);
  const trackData = useTrackData(trackSid) as LocalVideoTrackStats | RemoteVideoTrackStats | null;

  return (
    <>
      <Datum label="Dimensions" value={getDimensionString(dimensions)} />
      <Datum label="isSwitchedOff" value={String(isSwitchedOff)} />
      <Datum label="isEnabled" value={String(isEnabled)} />
      <Datum label="Bandwidth" value={String(trackBandwidth) + 'kbps'} />
      {trackData && (
        <>
          <Datum label="Codec" value={String(trackData.codec)} />
          <Datum label="Framerate" value={String(trackData?.frameRate)} />
          <Datum label="Packets Lost" value={String(trackData?.packetsLost)} />
        </>
      )}
    </>
  );
};

const Container = styled.div`
  &:not(:last-child) {
    border-bottom: 1px solid ${theme.borderColor};
    margin-bottom: 3px;
  }
  padding-bottom: 3px;

  &:hover {
    background: rgba(50, 50, 50, 0.9);
  }
`;

export const VideoTrackPublicationInfo: React.FC<{
  publication: LocalVideoTrackPublication | RemoteVideoTrackPublication;
}> = ({ publication }) => {
  const publishPriority = usePublishPriority(publication);
  const track = useTrack(publication) as LocalVideoTrack | RemoteVideoTrack | undefined;

  const handleMouseEnter = () => {
    // @ts-ignore
    track?._attachments?.forEach((el) => {
      el.style.filter = 'brightness(1.5)';
    });
  };

  const handleMouseLeave = () => {
    // @ts-ignore
    track?._attachments?.forEach((el) => {
      el.style.filter = '';
    });
  };

  return (
    <Container onMouseEnter={handleMouseEnter} onMouseLeave={handleMouseLeave}>
      <Datum label="Name" value={publication.trackName} />
      <Datum label="SID" value={publication.trackSid} />
      <Datum label="isSubscribed" value={String(!!track)} />
      <Datum label="publishPriority" value={String(publishPriority)} />
      {track && <VideoTrackInfo track={track} trackSid={publication.trackSid} />}
    </Container>
  );
};
