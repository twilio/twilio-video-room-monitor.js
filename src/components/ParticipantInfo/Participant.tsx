import React from 'react';
import { Accordion } from '../typography/Accordion/Accordion';
import Datum from '../typography/Datum/Datum';
import {
  Participant as ParticipantImpl,
  LocalVideoTrackPublication,
  RemoteVideoTrackPublication,
  LocalAudioTrackPublication,
  RemoteAudioTrackPublication,
} from 'twilio-video';
import useParticipantNetworkQualityLevel from '../../hooks/useParticipantNetworkQualityLevel/useParticipantNetworkQualityLevel';
import useParticipantIsReconnecting from '../../hooks/useParticipantIsReconnecting/useParticipantIsReconnecting';
import usePublications from '../../hooks/usePublications/usePublications';

import { VideoTrackPublicationInfo } from './VideoTrackPublicationInfo';
import { AudioTrackPublicationInfo } from './AudioTrackPublicationInfo';

export const Participant = ({ participant }: { participant: ParticipantImpl }) => {
  const networkQualityLevel = useParticipantNetworkQualityLevel(participant);
  const isReconnecting = useParticipantIsReconnecting(participant);

  const publications = usePublications(participant);

  const dataTrackPublications = publications.filter((publication) => publication.kind === 'data');
  const audioTrackPublications = publications.filter((publication) => publication.kind === 'audio');
  const videoTrackPublications = publications.filter((publication) => publication.kind === 'video');

  return (
    <Accordion label={participant.identity}>
      <Datum label="SID" value={participant.sid} />
      <Datum label="isReconnecting" value={String(isReconnecting)} />
      <Datum label="networkQualityLevel" value={String(networkQualityLevel)} />
      <Accordion label={`Data Tracks (${dataTrackPublications.length})`}>{/* TODO */}</Accordion>
      <Accordion label={`Audio Tracks (${audioTrackPublications.length})`}>
        {audioTrackPublications.map((audioTrackPublication) => (
          <AudioTrackPublicationInfo
            key={audioTrackPublication.trackSid}
            publication={audioTrackPublication as LocalAudioTrackPublication | RemoteAudioTrackPublication}
          />
        ))}
      </Accordion>
      <Accordion label={`Video Tracks (${videoTrackPublications.length})`}>
        {videoTrackPublications.map((videoTrackPublication) => (
          <VideoTrackPublicationInfo
            key={videoTrackPublication.trackSid}
            publication={videoTrackPublication as LocalVideoTrackPublication | RemoteVideoTrackPublication}
          />
        ))}
      </Accordion>
    </Accordion>
  );
};
