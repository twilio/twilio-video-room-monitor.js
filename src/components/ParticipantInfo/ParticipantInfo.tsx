import React from 'react';
import useParticipants from '../../hooks/useParticipants/useParticipants';
import useRoom from '../../hooks/useRoom/useRoom';
import useRoomState from '../../hooks/useRoomState/useRoomState';
import Headline from '../typography/common/Headline/Headline';
import { Participant } from './Participant/Participant';

export default function ParticipantInfo() {
  const participants = useParticipants();
  const room = useRoom();
  const roomState = useRoomState();

  if (!room || roomState === 'disconnected') return null;

  return (
    <>
      <Headline>Participant Information ({participants.length + 1}):</Headline>
      <Participant participant={room.localParticipant} isLocal={true} />
      {participants.map((participant) => (
        <Participant key={participant.sid} participant={participant} />
      ))}
    </>
  );
}
