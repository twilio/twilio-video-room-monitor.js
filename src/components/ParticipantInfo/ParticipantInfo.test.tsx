import React from 'react';
import { shallow } from 'enzyme';
import ParticipantInfo from './ParticipantInfo';
import useRoom from '../../hooks/useRoom/useRoom';

jest.mock('../../hooks/useParticipants/useParticipants', () => () => [
  {
    sid: 'testSid1',
    identity: 'test user1',
    isReconnecting: false,
    networkQualityLevel: 4,
  },
  {
    sid: 'testSid2',
    identity: 'test user2',
    isReconnecting: false,
    networkQualityLevel: 3,
  },
]);
jest.mock('../../hooks/useRoom/useRoom');

const mockUseRoom = useRoom as jest.Mock<any>;
mockUseRoom.mockImplementation(() => ({ room: {} }));

describe('the ParticipantInfo component', () => {
  it('should render correctly for each participant when a room exists', () => {
    const wrapper = shallow(<ParticipantInfo />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render anything when a room does not exist', () => {
    mockUseRoom.mockImplementationOnce(() => null);
    const wrapper = shallow(<ParticipantInfo />);
    expect(wrapper.getElement()).toBe(null);
  });
});
