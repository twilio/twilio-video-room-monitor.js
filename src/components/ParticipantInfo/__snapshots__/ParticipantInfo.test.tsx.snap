// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ParticipantInfo component should render correctly for each participant when a room exists 1`] = `
<Fragment>
  <styled.h4>
    Participant Information:
  </styled.h4>
  <Participant />
  <Participant
    key="testSid1"
    participant={
      Object {
        "identity": "test user1",
        "isReconnecting": false,
        "networkQualityLevel": 4,
        "sid": "testSid1",
      }
    }
  />
  <Participant
    key="testSid2"
    participant={
      Object {
        "identity": "test user2",
        "isReconnecting": false,
        "networkQualityLevel": 3,
        "sid": "testSid2",
      }
    }
  />
</Fragment>
`;
